src/ringmesh/io/geomodel_builder_resqml.cpp

```diff
@@ -0,0 +1,968 @@
+
+                builder_.geology.set_geological_entity_geol_feature(
+                    layer, GMGE::GEOL_FEATURE::STRATI_UNIT );
+
+                for( auto& region : geomodel_.regions() )
+                {
+                    builder_.geology.add_parent_children_relation(
+                        layer, region.gmme() );
+                }
+            }
+        }
+    }
+
+    namespace
+    {
+        bool read_property( AbstractValuesProperty& property,
+            const unsigned int patch_index,
+            GEO::AttributesManager& attri_manager )
+        {
+            if( property.getValuesHdfDatatype()
+                == AbstractValuesProperty::UNKNOWN )
+            {
+                throw RINGMeshException( "RESQML2", "Invalid property type: ",
+                    AbstractValuesProperty::UNKNOWN );
+            }
+
+            showAllMetadata( &property );
+            const unsigned int el_per_val = property.getElementCountPerValue();
+
+            const unsigned int val_count =
+                property.getValuesCountOfPatch( patch_index );
+
+            if( dynamic_cast< ContinuousProperty* >( &property ) != nullptr )
+            {
+                ContinuousProperty& continuousProp =
+                    static_cast< ContinuousProperty& >( property );
+
+                if( ( property.getValuesHdfDatatype()
+                            == AbstractValuesProperty::FLOAT
+                        || property.getValuesHdfDatatype()
+                               == AbstractValuesProperty::DOUBLE ) )
+                {
+                    GEO::Attribute< double > attribute;
+                    attribute.create_vector_attribute(
+                        attri_manager, property.getTitle(), el_per_val );
+
+                    std::unique_ptr< double[] > values( new double[val_count] );
+                    continuousProp.getDoubleValuesOfPatch(
+                        patch_index, &values[0] );
+                    for( auto val : range( val_count ) )
+                    {
+                        attribute[val] = values[val];
+                    }
+                }
+            }
+            else if( dynamic_cast< DiscreteProperty* >( &property ) != nullptr )
+            {
+                DiscreteProperty* discreteProp =
+                    dynamic_cast< DiscreteProperty* >( &property );
+
+                GEO::Attribute< int > attribute;
+                attribute.create_vector_attribute(
+                    attri_manager, property.getTitle(), el_per_val );
+
+                std::unique_ptr< int[] > values( new int[val_count] );
+                discreteProp->getIntValuesOfPatch( patch_index, &values[0] );
+                for( auto val : range( val_count ) )
+                {
+                    attribute[val] = values[val];
+                }
+            }
+
+            return true;
+        }
+    } // namespace
+
+    bool GeoModelBuilderRESQMLImpl::read_surfaces( const EpcDocument& pck )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this function to reduce its Cognitive Complexity from 30 to the 25 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3776)